### PR TITLE
Better support for CDATA

### PR DESCRIPTION
**Pull request by OrKoN**
- adds a share function to strip CDATA tags with decoding of entities outside of the tags
- adds a basic test

Open questions:
- not sure if it's a standard way to parse CDATA. Any ideas?
- anything else that this affects?

### CODE REVIEW

File: atom/parser.go [hunk 1]
```diff
@@ -681,10 +681,8 @@ func (ap *Parser) parseAtomText(p *xpp.XMLPullParser) (string, error) {
 	lowerType := strings.ToLower(text.Type)
 	lowerMode := strings.ToLower(text.Mode)
 
-	if strings.HasPrefix(result, "<![CDATA[") &&
-		strings.HasSuffix(result, "]]>") {
-		result = strings.TrimPrefix(result, "<![CDATA[")
-		result = strings.TrimSuffix(result, "]]>")
+	if strings.HasPrefix(result, "<![CDATA[") {
```
Comment (dy-dx) [inline atom/parser.go]: Should this use `strings.Contains` instead of `strings.HasPrefix`?

Comment (OrKoN) [reply atom/parser.go]: yep, makes sense. I will change it

**Comment by OrKoN** [comment]
It might make sense to introduce `shared.ContainsCDATA` because the check is in multiple places. Also, I am not sure if the CDATA parsing needs to depend on the content type.

**Comment by mmcdole** [comment]
@OrKoN thanks for your work on this.  This is definitely better than the current CDATA handling.

My original plan was a character-by-character parser of text elements which could handle CDATA elements without the need for contains() or indexof(), etc, and would have context to know when you were within an XML comment but I never got around to writing it.

However, this link has a lot of great examples/information on CDATA handling:
https://stackoverflow.com/questions/2784183/what-does-cdata-in-xml-mean

The following might make good testcases:

- Data inside the CDATA elements is not processed as XML characters
```xml
<exampleOfACDATA>
<![CDATA[
    Since this is a CDATA section
    I can use all sorts of reserved characters
    like > < " and &
    or write things like
    <foo></bar>
    but my document is still well formed!
]]>
</exampleOfACDATA>
```
or
```xml
<![CDATA[
Within this Character Data block I can
use double dashes as much as I want (along with <, &, ', and ")
*and* %MyParamEntity; will be expanded to the text
"Has been expanded" ... however, I can't use
the CEND sequence. If I need to use CEND I must escape one of the
brackets or the greater-than sign using concatenated CDATA sections.
]]>
```

- A comment following a CDATA element with CDATA end string (note that ]]> should not be stripped from the comment):
```xml
<!--
Within this comment I can use ]]>
and other reserved characters like <
&, ', and ", but %MyParamEntity; will not be expanded
(if I retrieve the text of this node it will contain
%MyParamEntity; and not "Has been expanded")
and I can't place two dashes next to each other.
-->
```

- Correct handling of escaped CENDs
```xml
<description>An example of escaped CENDs</description>
<!-- This text contains a CEND ]]> -->
<!-- In this first case we put the ]] at the end of the first CDATA block
     and the > in the second CDATA block -->
<data><![CDATA[This text contains a CEND ]]]]><![CDATA[>]]></data>
<!-- In this second case we put a ] at the end of the first CDATA block
     and the ]> in the second CDATA block -->
<alternative><![CDATA[This text contains a CEND ]]]><![CDATA[]>]]></alternative>
```

**Comment by OrKoN** [comment]
@mmcdole thanks for reviewing and the test cases. I have added them and it seems that the test case containing comment blocks is the only one that fails because the parser does not support comments (I didn't know there could be comments too) and `DecodeEntities` call fails on reserved characters. So comments are swallowed silently if they contain reserved characters.

Also, I am not sure that I got the expected results for test cases correctly. Perhaps some other is failing too.

Is there a way to re-use existing XML parsers to parse text nodes?

**Comment by mmcdole** [comment]
Merged this.  Better than what we have today. I’ll create a seperate issue to investigate leveraging standard library code or functions to enhance this further.

**Pull request merged**